src/prototype_room_paths.js

```diff
@@ -0,0 +1,184 @@
+};
+
+RoomPosition.prototype.getInternPath = function(target) {
+  let range;
+  if (_.isString(target)) {
+    range = 1;
+    target = Game.getObjectById(target).pos;
+  } else if (target instanceof RoomPosition) {
+    range = 0;
+  }
+  if (!target || target.room != this.room) {
+    return false;
+  }
+  console.log(JSON.stringify(target));
+  return PathFinder.search(
+    this, {
+      pos: target,
+      range: range,
+    }, {
+      swampCost: config.layout.swampCost,
+      plainCost: config.layout.plainCost
+    }
+  ).path;
+
+};
+
+Room.prototype.formatPath = function(path, pathName) {
+  let searchParent = (deep) => {
+    let paths = this.memory.paths;
+    let pathsNames = paths.list[deep];
+    let pathsAmount = pathsNames ? pathsNames.length : 0;
+    let p = 0;
+    console.log(deep, ' - - - ', JSON.stringify(pathsNames));
+    while (p < pathsAmount) {
+      if (_.eq(path[pos], paths[pathsNames[p]].path[0])) {
+        return pathsNames[p];
+      }
+      p++;
+    }
+    return false;
+  }
+
+  if (!path) {
+      return false;
+  }
+
+  let finish = false;
+  let parentEnd = false;
+  let deep = 0;
+  let pos = 0;
+  let parentPos;
+  let newPath = [];
+  let actualParentName;
+  let previousParentName;
+  let previousParentPos;
+  let paths = this.memory.paths;
+
+  while (!finish) {
+    actualParentName = searchParent(deep);
+
+    if (actualParentName) {
+      deep++;
+      parentPos = 0;
+      parentEnd = false;
+      while (!parentEnd) {
+        //console.log(JSON.stringify(paths[actualParentName].path[parentPos]));
+        //console.log(JSON.stringify(path[pos]));
+        if (!path[pos] || !_.eq(path[pos], paths[actualParentName].path[parentPos])) {
+          parentEnd = true;
+          previousParentName = actualParentName;
+        }
+        pos++;
+        parentPos++;
+
+      }
+    } else {
+      while (path[pos]) {
+        newPath.push(path[pos]);
+        pos++;
+      }
+      finish = true;
+    }
+  }
+
+  this.memory.paths.list[deep] = this.memory.paths.list[deep] || []
+  this.memory.paths.list[deep].push(pathName);
+  this.memory.paths[pathName] = { path: newPath, parentNode: [previousParentName, parentPos]};
+};
+
+Room.prototype.init = function() {
+  let controller = this.controller;
+  if (!controller) {
+    return false;
+  }
+  this.memory.paths = {list: []};
+  let exits = this.getExits();      // should return {E2: E2pos, E3: E3pos, E4: E4pos}
+  let sources = this.getSources();  // same
+  let mineral = this.getMineral();  // same
+  let n = 0;
+  let path;
+
+  while (n < 8) {
+      console.log(n);
+    if(n < 4) {
+      path = controller.pos.getInternPath(exits['E' + (n + 1)]);
+      this.formatPath(path, 'E' + (n + 1));
+    } else if (n < 7) {
+      path = controller.pos.getInternPath(sources['S' + (n - 3)]);
+      this.formatPath(path, 'S' + (n - 3));
+    } else {
+      path = controller.pos.getInternPath(mineral['M']);
+      this.formatPath(path, 'M');
+    }
```

Comment:
maybe add 

```let pathName;``` at L:168

```pathName = 'E' + (n + 1); ``` at L:173,  

```pathName = 'S' + (n - 3)``` at L:176 and

```pathName =  'M'``` at 179'  ... and here: 

``` this.formatPath(path, pathName);```

and maybe change L:174 to 
```
} 
if (n < 7) {
```